### PR TITLE
[inference] Add smmoothquant for llama

### PR DESCRIPTION
**Pull request by Xu-Kai**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number
 `closed #4902` `closed #4895`  `closed #4861` `closed #4854`  `closed #4853`  `closed #4850`  `closed #4843`

## 📝 What does this PR do?
- Add smoothquant.
- Support Llama model.
- Add save and load checkpoint for quantized model.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [x] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/inference/quant/smoothquant/models/base_model.py [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+    @abstractmethod
+    @torch.inference_mode()
+    def quantize(
+        self,
+        examples: List[Dict[str, Union[List[int], torch.LongTensor]]],
+    ):
+        if self.quantized:
+            raise EnvironmentError("can't execute quantize because the model is quantized.")
+
+    def to(self, device: Union[str, torch.device]):
```
Comment (ver217) [inline colossalai/inference/quant/smoothquant/models/base_model.py:104]: This method is redundant

File: colossalai/inference/quant/smoothquant/models/base_model.py [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+        init_contexts = [no_init_weights()]
+        if low_cpu_mem_usage:
+            init_contexts.append(accelerate.init_empty_weights(include_buffers=True))
+
+        with ContextManagers(init_contexts):
+            model = AutoModelForCausalLM.from_config(
+                config, trust_remote_code=trust_remote_code, torch_dtype=torch_dtype
+            )
+            if config.model_type == "llama":
+                from .llama import LlamaSmoothquantDecoderLayer, init_to_get_rotary, llama_model_forward
```
Comment (ver217) [inline colossalai/inference/quant/smoothquant/models/base_model.py:450]: You'd better do this in subclasses.

File: colossalai/kernel/cuda_native/csrc/smoothquant/linear.cu [hunk 3]
```diff
@@ -0,0 +1,162 @@
+// modified from https://github.com/Guangxuan-Xiao/torch-int/blob/main/torch_int/kernels/linear.cu
```
Comment (ver217) [inline colossalai/kernel/cuda_native/csrc/smoothquant/linear.cu:1]: Can you import this lib directly?

File: colossalai/inference/quant/smoothquant/models/llama.py [hunk 4]
```diff
@@ -0,0 +1,836 @@
...
+from transformers.utils import add_start_docstrings_to_model_forward
+
+from colossalai.kernel.triton import (
+    copy_kv_cache_to_dest,
+    int8_rotary_embedding_fwd,
+    smooth_llama_context_attn_fwd,
+    smooth_token_attention_fwd,
+)
+
+from ....tensor_parallel.batch_infer_state import BatchInferState
```
Comment (ver217) [inline colossalai/inference/quant/smoothquant/models/llama.py:33]: Don't use relative import for other folder

File: colossalai/kernel/triton/int8_rotary_embedding_kernel.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
+# Adapted from ModelTC https://github.com/ModelTC/lightllm
```
Comment (ver217) [inline colossalai/kernel/triton/int8_rotary_embedding_kernel.py:1]: Same problem. Can we import this lib directly?

File: tests/test_smoothquant/test_llama_attention.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    HAS_TRITON = False
+    print("please install triton from https://github.com/openai/triton")
+
+try:
+    from colossalai.inference.quant.smoothquant.models import LLamaSmoothquantAttention
+
+    HAS_TORCH_INT = True
+except ImportError:
+    HAS_TORCH_INT = False
+    print("Please install torch_int from https://github.com/Guangxuan-Xiao/torch-int")
```
Comment (ver217) [inline tests/test_smoothquant/test_llama_attention.py:19]: Do you need to update requirements-test.txt?

File: colossalai/kernel/cuda_native/csrc/smoothquant/linear.cu (hunk 3 above)
Comment (Xu-Kai) [reply colossalai/kernel/cuda_native/csrc/smoothquant/linear.cu:1]: The linear in torch int lib only support relu linear.  This linear file is to support silu linear.

File: colossalai/kernel/triton/int8_rotary_embedding_kernel.py (hunk 5 above)
Comment (Xu-Kai) [reply colossalai/kernel/triton/int8_rotary_embedding_kernel.py:1]: The method support int8 rortary embedding. The method in light lllm can't support.

File: colossalai/kernel/cuda_native/csrc/smoothquant/linear.cu (hunk 3 above)
Comment (ver217) [reply colossalai/kernel/cuda_native/csrc/smoothquant/linear.cu:1]: Then shall we add their license to ours? And do they allow modification?

File: colossalai/inference/quant/smoothquant/models/base_model.py (hunk 1 above)
Comment (Xu-Kai) [reply colossalai/inference/quant/smoothquant/models/base_model.py:104]: Removed.

File: colossalai/inference/quant/smoothquant/models/llama.py (hunk 4 above)
Comment (Xu-Kai) [reply colossalai/inference/quant/smoothquant/models/llama.py:33]: Updated.

File: tests/test_smoothquant/test_llama_attention.py (hunk 6 above)
Comment (Xu-Kai) [reply tests/test_smoothquant/test_llama_attention.py:19]: torch-int lib can only be installed from source.

File: colossalai/inference/quant/smoothquant/models/base_model.py (hunk 2 above)
Comment (Xu-Kai) [reply colossalai/inference/quant/smoothquant/models/base_model.py:450]: Updated.

**Comment by Xu-Kai** [comment]
image

File: colossalai/kernel/cuda_native/csrc/smoothquant/linear.cu (hunk 3 above)
Comment (Xu-Kai) [reply colossalai/kernel/cuda_native/csrc/smoothquant/linear.cu:1]: They use MIT License. I add the license for torch-in and smoothquant in the LICENSE

**Review by ver217 — APPROVED** [review]

**Review by tiandiao123 — APPROVED** [review]

**Pull request merged**